### PR TITLE
This cleans up the GET parameters sent to the game list page since it…

### PR DESCRIPTION
**Pull request by strycore**
… can cause Django to crash

### CODE REVIEW

**Comment by strycore** [comment]
Gonna merge this since I want to deploy it but you can still review it :)

File: games/views/pages.py [hunk 1]
```diff
@@ -67,11 +66,29 @@ def get(self, request, *args, **kwargs):
         }
         return super().get(request, *args, **kwargs)
 
+    def get_page(self):
+        """Return the page number for the paginated queryset"""
+        try:
+            page = int(self.request.GET.get("page"))
+        except ValueError:
+            return 1
+        if page < 1:
```
Comment (AlexanderRavenheart) [inline games/views/pages.py:75]: You also need to add a check if page > number of pages available.

File: games/views/pages.py [hunk 2]
```diff
@@ -67,11 +66,29 @@ def get(self, request, *args, **kwargs):
...
+            return int(self.request.GET.get("paginate_by", self.paginate_by))
+        except ValueError:
+            return self.paginate_by
 
     def get_ordering(self):
-        return self.request.GET.get('ordering', self.ordering)
+        """Return the field used to order the query by"""
+        field = self.request.GET.get("ordering", self.ordering)
+        if field.strip('-') not in Game.valid_fields():
+            return None
```
Comment (AlexanderRavenheart) [inline games/views/pages.py:90]: This should be return 'name'.

File: games/views/pages.py [hunk 3]
```diff
@@ -82,10 +99,12 @@ def get_queryset(self):
         queryset = queryset.prefetch_related(
             "genres", "publisher", "developer", "platforms", "installers"
         )
-        if self.q_params['q'] and not self.q_params['search-installers']:
-            queryset = queryset.order_by('-rank', self.get_ordering())
-        else:
-            queryset = queryset.order_by(self.get_ordering())
+        ordering = self.get_ordering()
+        if ordering:
```
Comment (AlexanderRavenheart) [inline games/views/pages.py:103]: Remove this, results should always be ordered to provide consistent behavior.

File: games/views/pages.py [hunk 4]
```diff
@@ -111,13 +130,27 @@ def get_filtered_queryset(self, queryset):
...
+                except AttributeError:
+                    continue
+                flag_q |= Q(flags=flag)
             queryset = queryset.filter(flag_q)
         return queryset
 
+    def clean_parameters(self):
+        """Validators used to prevent sending garbage data to Django views"""
+        # Set the GET QueryDict to mutable so we can corrrect the values
+        self.request.GET._mutable = True  # pylint: disable=protected-access
```
Comment (AlexanderRavenheart) [inline games/views/pages.py:145]: Why do you need to hack the GET dictionary?
In what scenario does paginate_by and page receive incorrect values?

File: games/tests/test_views.py [hunk 5]
```diff
@@ -44,10 +44,49 @@ def test_can_access_installer_feed(self):
...
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200)
+        url = "/games?page=-1"
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200)
+        url = "/games?page=0"
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200)
+
+    def test_can_receive_garbage_pagination_by(self):
```
Comment (AlexanderRavenheart) [inline games/tests/test_views.py:79]: Invalid test. When can paginate_by be invalid? It is not a user controlled dynamic parameter, but a set value from a predefined list of values.

File: games/views/pages.py (hunk 4 above)
Comment (strycore) [reply games/views/pages.py:145]: users can set them in the URL. This generates 500 errors, I receive 1 email for each of those.

File: games/tests/test_views.py (hunk 5 above)
Comment (strycore) [reply games/tests/test_views.py:79]: `yes this is a user controlled value.

File: games/views/pages.py (hunk 1 above)
Comment (strycore) [reply games/views/pages.py:75]: unneeded unless it can crash the website

File: games/views/pages.py (hunk 3 above)
Comment (strycore) [reply games/views/pages.py:103]: not really important

File: games/views/pages.py (hunk 2 above)
Comment (strycore) [reply games/views/pages.py:90]: this is an arbitrary decision, it can work both ways

**Pull request merged**